rand_distr/src/hypergeometric.rs

```diff
@@ -0,0 +1,404 @@
+/// `f(k) = binomial(K, k) * binomial(N-K, n-k) / binomial(N, n)`,
+/// where `binomial(a, b) = a! / (b! * (a - b)!)`.
+/// 
+/// The [binomial distribution](crate::Binomial) is the analagous distribution
+/// for sampling with replacement. It is a good approximation when the population
+/// size is much larger than the sample size.
+/// 
+/// # Example
+/// 
+/// ```
+/// use rand_distr::{Distribution, Hypergeometric};
+///
+/// let hypergeo = Hypergeometric::new(60, 24, 7).unwrap();
+/// let v = hypergeo.sample(&mut rand::thread_rng());
+/// println!("{} is from a hypergeometric distribution", v);
+/// ```
+#[derive(Copy, Clone, Debug)]
+pub struct Hypergeometric {
+    n1: u64,
+    n2: u64,
+    k: u64,
+    offset_x: i64,
+    sign_x: i64,
+    sampling_method: SamplingMethod,
+}
+
+/// Error type returned from `Hypergeometric::new`.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Error {
+    /// `total_population_size` is too large, causing floating point underflow.
+    PopulationTooLarge,
+    /// `population_with_feature > total_population_size`.
+    ProbabilityTooLarge,
+    /// `sample_size > total_population_size`.
+    SampleSizeTooLarge,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(match self {
+            Error::PopulationTooLarge => "total_population_size is too large causing underflow in geometric distribution",
+            Error::ProbabilityTooLarge => "population_with_feature > total_population_size in geometric distribution",
+            Error::SampleSizeTooLarge => "sample_size > total_population_size in geometric distribution",
+        })
+    }
+}
+
+// evaluate fact(numerator.0)*fact(numerator.1) / fact(denominator.0)*fact(denominator.1)
+fn fraction_of_products_of_factorials(numerator: (u64, u64), denominator: (u64, u64)) -> f64 {
+    let min_top = u64::min(numerator.0, numerator.1);
+    let min_bottom = u64::min(denominator.0, denominator.1);
+    // the factorial of this will cancel out:
+    let min_all = u64::min(min_top, min_bottom);
+
+    let max_top = u64::max(numerator.0, numerator.1);
+    let max_bottom = u64::max(denominator.0, denominator.1);
+    let max_all = u64::max(max_top, max_bottom);
+
+    let mut result = 1.0;
+    for i in (min_all + 1)..=max_all {
+        if i <= min_top {
+            result *= i as f64;
+        }
+        
+        if i <= min_bottom {
+            result /= i as f64;
+        }
+        
+        if i <= max_top {
+            result *= i as f64;
+        }
+        
+        if i <= max_bottom {
+            result /= i as f64;
+        }
+    }
+    
+    result
+}
+
+fn ln_of_factorial(v: f64) -> f64 {
+    // the paper calls for ln(v!), but also wants to pass in fractions,
+    // so we need to use Stirling's approximation to fill in the gaps:
+    v * v.ln() - v
+}
+
+impl Hypergeometric {
+    /// Constructs a new `Hypergeometric` with the given shape parameters.
```

Comment:
For clarity, it would be nice if you could document how those relate to the variables in the definition of the PDF above.